src/GafferVDB/VDBObject.cpp

```diff
@@ -0,0 +1,293 @@
+	auto& v = newByteArray->writable();
+	v.resize( s.length() );
+	std::copy( s.begin(), s.end(), v.begin() );
+
+	return newByteArray;
+}
+
+void VDBObject::forceRead(const std::string& name)
+{
+	openvdb::GridBase::Ptr grid = findGrid( name );
+	grid->readNonresidentBuffers();
+}
+
+IECore::CompoundObjectPtr VDBObject::metadata(const std::string& name) const
+{
+	//todo dirty this when the grid has been updated rather
+	//than assuming it always has to be done.
+	openvdb::GridBase::ConstPtr grid = findGrid( name );
+
+	if (!grid)
+	{
+		return IECore::CompoundObjectPtr();
+	}
+
+	CompoundObjectPtr metadata = new CompoundObject();
+
+	for (auto metaIt = grid->beginMeta(); metaIt != grid->endMeta(); ++metaIt)
+	{
+		openvdb::Metadata::Ptr ptr = metaIt->second;
+
+		if (metaIt->second->typeName() == "string")
+		{
+			openvdb::TypedMetadata<openvdb::Name>::ConstPtr typedPtr = openvdb::DynamicPtrCast<openvdb::TypedMetadata<openvdb::Name> >(ptr);
+
+			if (typedPtr)
+			{
+				StringDataPtr stringData = new StringData();
+				stringData->writable() = typedPtr->value();
+				metadata->members()[metaIt->first] = stringData;
+			}
+		}
+		else if (metaIt->second->typeName() == "int64")
+		{
+			openvdb::TypedMetadata<openvdb::Int64>::ConstPtr typedPtr = openvdb::DynamicPtrCast<openvdb::TypedMetadata<openvdb::Int64> >(ptr);
+			if (typedPtr)
+			{
+				Int64DataPtr intData = new Int64Data();
+				intData->writable() = typedPtr->value();
+				metadata->members()[metaIt->first] = intData;
+			}
+		}
+		else if (metaIt->second->typeName() == "bool")
+		{
+			openvdb::TypedMetadata<bool>::ConstPtr typedPtr = openvdb::DynamicPtrCast<openvdb::TypedMetadata<bool> > (ptr);
+			if (typedPtr)
+			{
+				BoolDataPtr data = new BoolData();
+				data->writable() = typedPtr->value();
+				metadata->members()[metaIt->first] = data;
+			}
+
+		}
+		else if (metaIt->second->typeName() == "vec3i")
+		{
+			openvdb::TypedMetadata<openvdb::math::Vec3i>::ConstPtr typedPtr = openvdb::DynamicPtrCast<openvdb::TypedMetadata<openvdb::math::Vec3i> >(ptr);
+			if (typedPtr)
+			{
+				V3iDataPtr data = new V3iData();
+				data->writable() = Imath::V3i( typedPtr->value().x(), typedPtr->value().y(), typedPtr->value().z() );
+				metadata->members()[metaIt->first] = data;
+			}
+		}
+		else
+		{
+			StringDataPtr stringData = new StringData();
+			stringData->writable() = "unsupported type: " + metaIt->second->typeName();
+			metadata->members()[metaIt->first] = stringData;
+		}
+	}
+	return metadata;
+}
+
+
+bool VDBObject::isEqualTo( const IECore::Object *other ) const
+{
+	if (!IECore::VisibleRenderable::isNotEqualTo( other ))
+	{
+		return false;
+	}
+
+	const VDBObject *vdbObject = runTimeCast<const VDBObject>(other);
+
+	// are the grid pointers identical
+	return *m_grids == *vdbObject->m_grids;
+}
+
+void VDBObject::hash( IECore::MurmurHash &h ) const
+{
+	IECore::VisibleRenderable::hash( h );
+
+	// hash the pointers of the grids for now. todo discuss with John / Andrew
```

Comment:
Bump

Reply:
I was concerned about the performance & ensuring we capture all important data in 'data' hash of VDBs. I really don't like this pointer hash so how about the following: 

If we have a VDBGrid type then we can store the hash for the grid and hash these in the VDBObject? Using a Murmur hash stream on a grid would allow us to ensure we capture all important data, what do you think?

I'm leaning towards a VDBGrid type which will allow use to cache the hash.

Reply:
> I was concerned about the performance & ensuring we capture all important data in 'data' hash of VDBs. Using a Murmur hash stream on a grid would allow us to ensure we capture all important data

Ah, OK, I wasn't sure how we'd do the hashing but a stream-based approach sounds like it'd work. Presumably we could have a special case for when we know we've just loaded the grids from file, and could just hash the filename in that case?

> If we have a VDBGrid type then we can store the hash for the grid and hash these in the VDBObject?

Without some details about what you propose, it's not clear to me if this is the way to go or not. If VDBGrid is a full Object derived class (I think you had something like this in a previous iteration) then it feels like it might be a bit excessive - as I keep saying, I think VDBObject should be as minimal a wrapper as possible. If we can do what we need without an extra layer of (public) indirection, then that would be my preference. I suggest we chat about this in tomorrow's meeting, but so we have something concrete to talk about, here's a proposal at the simple end of the scale :

```
VDBObject
{

    public :

        // Accessors pretty much as you have them already

    private :

        struct HashedGrid
        {
            // The actual grid, and the only thing we ever
            // expose publicly.
            openvdb::GridPtr grid;
            // Set false by `insertGrid()` and the non-const
            // version of `findGrid()`
            bool hashValid;
            // Updated lazily as needed, based on the hashValid
            // field.
            IECore::MurmurHash hash;
        };

        // Internal storage for grids and their hashes. We use
        // shared_ptr so that `copyFrom()` can do COW, querying
        // `shared_ptr::unique()` to perform the
        // lazy copy in the non-const version of `findGrid()`.
        typedef std::shared_ptr<HashedGrid> HashedGridPtr;
        std::unordered_map<std::string, HashedGridPtr> m_grids;

};
```